docs/development/design/userspace-mode.md

```diff
@@ -0,0 +1,108 @@
+
+To support a wider set of use cases in Nexodus, the proposal is to allow `nexd` to run in a new userspace-only mode where it can act as a proxy between locally accessible applications and other devices accessible via the Wireguard network managed by Nexodus. This turns `nexd` into a layer 4 (transport) interface to a Nexodus network instead of layer 3 (IP).
+
+### CLI Subcommands
+
+To allow `nexd` to be run in a different mode, `nexd`'s command line interface shold be modified to make use of subcommands. This will allow a different set of flags to be defined for each operating mode. The first step is to move all existing flags under a new subcommand, `agent`.
+
+```sh
+sudo nexd agent https://try.nexodus.local
+```
+
+While the implementation is not considered in scope of this proposal, it should be considered a natural next step to change the `--relay-node` and `--discovery-node` flags into subcommands. For example:
+
+```sh
+sudo nexd relay https://try.nexodus.local
+```
+
+or
+
+```sh
+sudo nexd discovery https://try.nexodus.local
+```
+
+Once the current mode has been moved under an `agent` subcommand, a new `proxy` command can be added for this new userspace-only mode of interfacing with a Nexodus network.
+
+## Fewer Flags for Proxy Mode
+
+Only a subset of the flags used in `agent` mode apply to `proxy` mode. The following flags will not be present under the `proxy` mode:
+
+```text
+   --child-prefix value [ --child-prefix value ]  Request a CIDR range of addresses that will be advertised from this node (optional) [$NEXD_REQUESTED_CHILD_PREFIX]
+   --relay-node                                   Set if this node is to be the relay node for a hub and spoke scenarios (default: false) [$NEXD_RELAY_NODE]
+   --discovery-node                               Set if this node is to be the discovery node for NAT traversal in an organization (default: false) [$NEXD_DISCOVERY_NODE]
+   --relay-only                                   Set if this node is unable to NAT hole punch in a hub zone (Nexodus will set this automatically if symmetric NAT is detected) (default: false) [$NEXD_RELAY_ONLY]
+```
+
+### Proxy Configuration
+
+As a first step, `nexd` will take proxy configuration via two configuration flags: `--ingress` and `--egress`.
+
+`--ingress` is used to proxy for connections coming from the Nexodus network to a server accessible by `nexd` via a local network.
+
+`--egress` is used to proxy for connectoins coming from a network local to `nexd` to a server reachable over the Nexodus network.
+
+In both cases, the format for the value is `[protocol]:[port]:[desination_ip]:[destination_port]`. `protocol` can be `tcp` or `udp` to start.
+
```

Comment:
At some point it might be nice to have `nextctl` commands to manage the ingress/egress list.  Kinda like:
* `nexctl list [ingress|egress]`
* `nexctl [create|delete] [ingress|egress] <port-mapping>`

This would be super handy if your on a system and and the services you want to access or or are exposing a bit dynamic and you don't want to have interrupt existing connections.  

If we do that, would those persist between nexd restarts?   If they do, does it kind of ignore options on subsequent restarts of the `nexd` cli?

Reply:
Yeah, I agree that runtime management would be useful. See a few paragraphs down for some comments on this. I proposed leaving for the future because I didn’t want to deal with it yet. 😀